### DISCUSSION TITLE
RTSP Display Server

### DISCUSSION THREAD

**Discussion by Perek101** [pedroSG94/RootEncoder#1342] [Q&A] [answered]
Hello, Pedro! I noticed the RTSP-Server in your repositories. Among the examples, I only found streaming from the camera, but in the library, I saw the RtspServerDisplay class. Do you have an example for this class or any guidance on how to test it? Thank you.

**Comment by pedroSG94** [comment]
Hello,

That class usage is really similar to this example:
https://github.com/pedroSG94/RootEncoder/tree/master/app/src/main/java/com/pedro/streamer/displayexample

You only need use RtspServerDisplay and use startStream method without string parameter like camera example:
https://github.com/pedroSG94/RTSP-Server/blob/master/app/src/main/java/com/pedro/sample/CameraDemoActivity.kt#L102

**Reply by Perek101** [reply]
I tried to do as you said, but an error occurred. Can you suggest what might be the issue?
```
/**
 * More documentation see:
 * [com.pedro.library.base.DisplayBase]
 * [com.pedro.library.rtmp.RtmpDisplay]
 */
@RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
class DisplayActivity : AppCompatActivity(), View.OnClickListener, ConnectCheckerRtsp, ConnectCheckerRtmp {

    private lateinit var rtspServerCamera1: RtspServerDisplay
    private lateinit var surfaceView: SurfaceView
    private lateinit var button: Button
    private lateinit var etUrl: EditText
    private val REQUEST_CODE_STREAM = 179 //random num
    private val REQUEST_CODE_RECORD = 180 //random num

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        setContentView(R.layout.activity_display)
        button = findViewById(R.id.b_start_stop)
        button.setOnClickListener(this)
        etUrl = findViewById(R.id.et_rtp_url)
        etUrl.hint = getString(R.string.hint_rtmp)
        val displayService = DisplayService.INSTANCE
        rtspServerCamera1 = RtspServerDisplay(this, false, this, 1935)
        //No streaming/recording start service
        if (displayService == null) {
            startService(Intent(this, DisplayService::class.java))
        }
        if (displayService != null && displayService.isStreaming()) {
            button.setText(R.string.stop_button)
        } else {
            button.setText(R.string.start_button)
        }
    }


    override fun onDestroy() {
        super.onDestroy()
        val displayService = DisplayService.INSTANCE
        if (displayService != null && !displayService.isStreaming() && !displayService.isRecording()) {
            //stop service only if no streaming or recording
            stopService(Intent(this, DisplayService::class.java))
        }
    }

    override fun onConnectionStartedRtmp(rtmpUrl: String) {}

    override fun onConnectionSuccessRtmp() {
        Toast.makeText(this@DisplayActivity, "Connection success", Toast.LENGTH_SHORT).show()
    }

    override fun onConnectionFailedRtmp(reason: String) {
        Toast.makeText(this@DisplayActivity, "Connection failed. $reason", Toast.LENGTH_SHORT).show()
        val displayService = DisplayService.INSTANCE
        if (displayService != null) {
            rtspServerCamera1.stopStream()
        }
        button.setText(R.string.start_button)
    }

    override fun onNewBitrateRtmp(bitrate: Long) {}

    override fun onDisconnectRtmp() {
        Toast.makeText(this@DisplayActivity, "Disconnected", Toast.LENGTH_SHORT).show()
    }

    override fun onAuthErrorRtmp() {
        Toast.makeText(this@DisplayActivity, "Auth error", Toast.LENGTH_SHORT).show()
    }

    override fun onAuthSuccessRtmp() {
        Toast.makeText(this@DisplayActivity, "Auth success", Toast.LENGTH_SHORT).show()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (data != null && (requestCode == REQUEST_CODE_STREAM
                    || requestCode == REQUEST_CODE_RECORD && resultCode == Activity.RESULT_OK)
        ) {
            val displayService = DisplayService.INSTANCE
            if (displayService != null) {
                val endpoint = etUrl.text.toString()
                displayService.prepareStreamRtp(endpoint, resultCode, data)

                displayService.startStreamRtp(endpoint)

            }

        } else {
            Toast.makeText(this, "No permissions available", Toast.LENGTH_SHORT).show()
            button.setText(R.string.start_button)
        }
    }


    override fun onClick(view: View) {
        val displayService = DisplayService.INSTANCE
        if (displayService != null) {
            if (view.id == R.id.b_start_stop) {
                if (!displayService.isStreaming()) {
                    button.setText(R.string.stop_button)

                    startActivityForResult(displayService.sendIntent()!!, REQUEST_CODE_STREAM)
                    rtspServerCamera1.startStream()

                } else {
                    button.setText(R.string.start_button)
                    displayService.stopStream()
                }

            }
        }
    }

    override fun onAuthErrorRtsp() {

    }

    override fun onAuthSuccessRtsp() {

    }

    override fun onConnectionFailedRtsp(reason: String) {

    }

    override fun onConnectionStartedRtsp(rtspUrl: String) {

    }

    override fun onConnectionSuccessRtsp() {

    }

    override fun onDisconnectRtsp() {

    }

    override fun onNewBitrateRtsp(bitrate: Long) {

    }
}
```
I have error:
FATAL EXCEPTION: main
Process: com.pedro.sample, PID: 31899
java.lang.RuntimeException: You need send intent data before startRecord or startStream
  at com.pedro.library.base.DisplayBase.startEncoders(DisplayBase.java:397)
  at com.pedro.library.base.DisplayBase.startStream(DisplayBase.java:388)
  at com.pedro.rtspserver.RtspServerDisplay.startStream(RtspServerDisplay.kt:36)
  at com.pedro.sample.DisplayActivity.onClick(DisplayActivity.kt:131)
  at android.view.View.performClick(View.java:7125)
  at android.view.View.performClickInternal(View.java:7102)
  at android.view.View.access$3500(View.java:801)
  at android.view.View$PerformClick.run(View.java:27336)
  at android.os.Handler.handleCallback(Handler.java:883)
  at android.os.Handler.dispatchMessage(Handler.java:100)
  at android.os.Looper.loop(Looper.java:214)
  at android.app.ActivityThread.main(ActivityThread.java:7356)
  at java.lang.reflect.Method.invoke(Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)

**Reply by Perek101** [reply]
Perhaps I am starting the stream incorrectly. How should I do it? Currently, I am doing it like this:
rtspServerCamera1 = RtspServerDisplay(this, false, this, 1935)
if (rtspServerCamera1.isRecording || rtspServerCamera1.prepareAudio() && rtspServerCamera1.prepareVideo()) {
      rtspServerCamera1.startStream()
    }

**Reply by pedroSG94** [reply]
Can you show me the code on displayService.prepareStreamRtp(endpoint, resultCode, data)?
That method should call sendIntent like this:
https://github.com/pedroSG94/RootEncoder/blob/master/app/src/main/java/com/pedro/streamer/displayexample/DisplayService.kt#L157

**Reply by Perek101** [reply]
I'm using your class DisplayService
fun prepareStreamRtp(endpoint: String, resultCode: Int, data: Intent) {
    Log.d("data intent", data.toString())
    keepAliveTrick()
    stopStream()
    if (endpoint.startsWith("rtmp")) {
      displayBase = RtmpDisplay(baseContext, true, connectCheckerRtp)
      displayBase?.setIntentResult(resultCode, data)
    } else {
      displayBase = RtspDisplay(baseContext, true, connectCheckerRtp)
      displayBase?.setIntentResult(resultCode, data)
    }
    displayBase?.glInterface?.setForceRender(true)
  }

  fun startStreamRtp(endpoint: String) {
    if (displayBase?.isStreaming != true) {
      if (displayBase?.prepareVideo() == true && displayBase?.prepareAudio() == true) {
        displayBase?.startStream(endpoint)
      }
    } else {
      showNotification("You are already streaming :(")
    }
  }
  
  I noticed that I have resultCode = -1 in onActivityResult

**Reply by pedroSG94** [reply]
Make sure that you have permissions added(Just add all for now):
https://github.com/pedroSG94/RootEncoder/blob/master/app/src/main/AndroidManifest.xml
And service properly declared:
https://github.com/pedroSG94/RootEncoder/blob/master/app/src/main/AndroidManifest.xml#L139

**Reply by Perek101** [reply]
I'm already adding all the permissions and display service. I found an error (should I have added ```rtspServerCamera1?.setIntentResult(resultCode, data)``` before `startStream`), thanks for the help

**Reply by pedroSG94** [reply, marked as answer]
Yes, that method should be before startStream and before prepareVideo/prepareAudio
